tests/components/tasmota/test_sensor.py

```diff
@@ -0,0 +1,254 @@
+)
+
+from tests.async_mock import patch
+from tests.common import async_fire_mqtt_message
+
+DEFAULT_SENSOR_CONFIG = {
+    "sn": {
+        "Time": "2020-09-25T12:47:15",
+        "DHT11": {"Temperature": None},
+        "TempUnit": "C",
+    }
+}
+
+
+async def test_controlling_state_via_mqtt(hass, mqtt_mock, setup_tasmota):
+    """Test state update via MQTT."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    sensor_config = copy.deepcopy(DEFAULT_SENSOR_CONFIG)
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/sensors",
+        json.dumps(sensor_config),
+    )
+    await hass.async_block_till_done()
+
+    state = hass.states.get("sensor.dht11_temperature")
+    assert state.state == "unavailable"
+    assert not state.attributes.get(ATTR_ASSUMED_STATE)
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/LWT", "Online")
+    state = hass.states.get("sensor.dht11_temperature")
+    assert state.state == STATE_UNKNOWN
+    assert not state.attributes.get(ATTR_ASSUMED_STATE)
+
+    # Test periodic state update
+    async_fire_mqtt_message(
+        hass, "tasmota_49A3BC/tele/SENSOR", '{"DHT11":{"Temperature":20.5}}'
+    )
+    state = hass.states.get("sensor.dht11_temperature")
+    assert state.state == "20.5"
+
+    # Test polled state update
+    async_fire_mqtt_message(
+        hass,
+        "tasmota_49A3BC/stat/STATUS8",
+        '{"StatusSNS":{"DHT11":{"Temperature":20.0}}}',
+    )
+    state = hass.states.get("sensor.dht11_temperature")
+    assert state.state == "20.0"
+
+
+async def test_attributes(hass, mqtt_mock, setup_tasmota):
+    """Test correct attributes for sensors."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    sensor_config = {
+        "sn": {
+            "DHT11": {"Temperature": None},
+            "Beer": {"CarbonDioxide": None},
+            "TempUnit": "C",
+        }
+    }
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/sensors",
+        json.dumps(sensor_config),
+    )
+    await hass.async_block_till_done()
+
+    state = hass.states.get("sensor.dht11_temperature")
+    assert state.attributes.get("device_class") == "temperature"
+    assert state.attributes.get("friendly_name") == "DHT11 Temperature"
+    assert state.attributes.get("icon") is None
+    assert state.attributes.get("unit_of_measurement") == "C"
+
+    state = hass.states.get("sensor.beer_CarbonDioxide")
+    assert state.attributes.get("device_class") is None
+    assert state.attributes.get("friendly_name") == "Beer CarbonDioxide"
+    assert state.attributes.get("icon") == "mdi:molecule-co2"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6319846)